src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -611,22 +613,86 @@ internal static bool WasServerMutexOpen(string mutexName)
         {
             try
             {
-                Mutex mutex;
-                var open = Mutex.TryOpenExisting(mutexName, out mutex);
-                if (open)
-                {
-                    mutex.Dispose();
-                    return true;
-                }
+                var open = DoesMutexExist(mutexName);
+                return open;
             }
             catch
             {
                 // In the case an exception occured trying to open the Mutex then 
                 // the assumption is that it's not open. 
                 return false;
             }
+        }
 
-            return false;
+        public static bool DoesMutexExist(string mutexName) 
+        {
+            if (PlatformInformation.IsRunningOnMono) 
+            {
+                bool createdNew;
+                using (var temp = new ServerFileMutexPair(mutexName, out createdNew))
+                    return !createdNew;
+            } 
+            else 
+            {
+                Mutex temp;
+                var result = Mutex.TryOpenExisting(mutexName, out temp);
+                if (result)
+                    temp.Dispose();
+                return result;
+            }
+        }
+
+        public static bool WaitMutex(IDisposable mutex, int millisecondsTimeout) 
+        {
+            if (PlatformInformation.IsRunningOnMono) 
+            {
+                if (millisecondsTimeout != 0)
+                    throw new NotImplementedException("Wait on mutex");
+
+                var fm = (ServerFileMutexPair)mutex;
+                var acquiredLock = fm.Lock();
+                if (acquiredLock)
+                    fm.Unlock();
+                return acquiredLock;
+            } 
+            else 
+            {
+                var temp = (Mutex)mutex;
+                return temp.WaitOne(millisecondsTimeout: millisecondsTimeout);
+            }
+        }
+
+        public static void ReleaseMutex(IDisposable mutex) 
+        {
+            if (PlatformInformation.IsRunningOnMono) 
+            {
+                var fm = (ServerFileMutexPair)mutex;
+                fm.Unlock();
+            }
+            else 
+            {
+                var temp = (Mutex)mutex;
```

Comment:
All the casting here makes this feel a bit fragile. Think we should consider just abstracting this away. Example consider if we had `OpenOrCreateMutex` create a type like this: 

```
internal abstract class ServerMutex : IDisposable {
   internal abstract Release();
   internal abstract Wait(TimeSpan timeSpan);
   internal static ServerMutex OpenOrCreate(bool initiallyOwned, string name, bool createdNew);
}
```

Then have a Mono and non-Mono implementation.